internal/rotate_est.go

```diff
@@ -0,0 +1,178 @@
+package internal
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"go.mozilla.org/pkcs7"
+)
+
+var (
+	oidKeyUsage         = asn1.ObjectIdentifier([]int{2, 5, 29, 15})
+	oidExtendedKeyUsage = asn1.ObjectIdentifier([]int{2, 5, 29, 37})
+
+	asn1DigitalSignature = []byte{3, 2, 7, 128}
+	asn1TlsWebClientAuth = []byte{48, 10, 6, 8, 43, 6, 1, 5, 5, 7, 3, 2}
+)
+
+type estStep struct{}
+
+func (s estStep) Name() string {
+	return "Generate new certificate"
+}
+
+func (s estStep) Completed(state CertRotationState) bool {
+	return len(state.NewKey) > 0 && len(state.NewCert) > 0
+}
+
+func (s estStep) Execute(handler *CertRotationHandler) error {
+	// Find the current certificate so we can build the proper EST request
+	tlsCert := handler.client.Transport.(*http.Transport).TLSClientConfig.Certificates[0]
+	cert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return nil
+	}
+
+	var signer crypto.Signer
+	var newKey string
+
+	// Generate a new private key
+	if handler.usePkcs11() {
+		newKey = "07"
+		signer, err = handler.crypto.ctx.GenerateECDSAKeyPair(idToBytes(newKey), elliptic.P256())
```

Comment:
Shouldn't the key type be configurable so a user can choose?

Reply:
I don't see a reason to. We should build fioconfig with the best option that interoperates with our DG.